packages/button/index.tsx

```diff
@@ -0,0 +1,99 @@
+// The MIT License
+//
+// Copyright (c) 2018 Google, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import * as React from 'react';
+import * as classnames from 'classnames';
+// @ts-ignore
+import withRipple from '@material/react-ripple';
+
+export interface ButtonProps<T> {
+  raised?: boolean,
+  unelevated?: boolean,
+  outlined?: boolean,
+  dense?: boolean,
+  disabled?: boolean,
+  unbounded?: boolean,
+  initRipple?: (surface: T) => void,
```

Comment:
Should this be `React.Ref<T>`?

Reply:
No, we actually need a domElement. MDC Web requires it.

Reply:
You ought to be able to get a DOM element out of a ref.  All I see here is that you're passing a ref to a `react-dom` element, so the type should probably be `Ref<T>`.  Feel free to come explain it to me if there's something I'm missing.

Reply:
You mean this `initRipple?: (surface: React.Ref<T>) => void;`?

T is the `HTMLButtonElement` or `HTMLAnchorElement`, not a Ref. The React.Ref definition is

```
type Ref<T> = { bivarianceHack(instance: T | null): void }["bivarianceHack"] | RefObject<T> | null;```

It is not a RefObject or null. It is the `instance T`. So I'm not sure why just definite it as `T` is ok. `surface` is being passed through to the `initRipple` function, which is not a `react-dom` element. So not sure what you mean there either.

I can chat with you about this tomorrow.

Also I get this error:

<img width="1637" alt="screen shot 2018-12-18 at 00 17 15" src="https://user-images.githubusercontent.com/579873/50140709-7f71fe00-025a-11e9-88b4-ec0ebf33e405.png">

Reply:
I just cloned your repo and tried it.  I had no problems:

```typescript
export interface ButtonProps<T> {
  raised?: boolean;
  unelevated?: boolean;
  outlined?: boolean;
  dense?: boolean;
  disabled?: boolean;
  unbounded?: boolean;
  initRipple?: React.Ref<T>;
  className?: string;
  icon?: React.ReactElement<React.HTMLProps<HTMLOrSVGElement>>;
  href?: string;
}
```

It built cleanly:

```
~/Projects/material-components-web-react/packages/button ☂ feat/typescript-button
❯ ../../node_modules/.bin/tsc ./index.tsx --jsx react --lib ESNext,DOM
```

The only in-editor error I see is a complaint that you can't `...otherProps`: "Rest types may only be created from object types."

Reply:
Super strange. TS compiler was barfing earlier. This is the 3rd time i tried it. Pushing that change.